arbor/include/arbor/event_generator.hpp

```diff
@@ -49,130 +50,38 @@ namespace arb {
-//    function before the first call to the `events` method.
+// `event_generator` objects have value semantics.
 
 using event_seq = std::pair<const spike_event*, const spike_event*>;
 using resolution_function = std::function<cell_lid_type(const cell_local_label_type&)>;
 
-// The simplest possible generator that generates no events.
-// Declared ahead of event_generator so that it can be used as the default
-// generator.
-struct empty_generator {
-    void reset() {}
-    event_seq events(time_type, time_type) {
-        return {nullptr, nullptr};
-    }
-    void resolve_label(resolution_function) {}
-};
-
-class event_generator {
-public:
-    event_generator(): event_generator(empty_generator()) {}
-
-    template <typename Impl, std::enable_if_t<!std::is_same<std::decay_t<Impl>, event_generator>::value, int> = 0>
-    event_generator(Impl&& impl):
-        impl_(new wrap<Impl>(std::forward<Impl>(impl)))
-    {}
-
-    event_generator(event_generator&& other) = default;
-    event_generator& operator=(event_generator&& other) = default;
-
-    event_generator(const event_generator& other):
-        impl_(other.impl_->clone())
-    {}
-
-    event_generator& operator=(const event_generator& other) {
-        impl_ = other.impl_->clone();
-        return *this;
-    }
-
-    void reset() {
-        impl_->reset();
-    }
-
-    event_seq events(time_type t0, time_type t1) {
-        return impl_->events(t0, t1);
-    }
-
-    void resolve_label(resolution_function label_resolver) {
-        impl_->resolve_label(std::move(label_resolver));
-    }
-
-private:
-    struct interface {
-        virtual void reset() = 0;
-        virtual void resolve_label(resolution_function) = 0;
-        virtual event_seq events(time_type, time_type) = 0;
-        virtual std::unique_ptr<interface> clone() = 0;
-        virtual ~interface() {}
-    };
-
-    std::unique_ptr<interface> impl_;
-
-    template <typename Impl>
-    struct wrap: interface {
-        explicit wrap(const Impl& impl): wrapped(impl) {}
-        explicit wrap(Impl&& impl): wrapped(std::move(impl)) {}
-
-        event_seq events(time_type t0, time_type t1) override {
-            return wrapped.events(t0, t1);
-        }
-
-        void reset() override {
-            wrapped.reset();
-        }
-
-        void resolve_label(resolution_function label_resolver) override {
-            wrapped.resolve_label(std::move(label_resolver));
-        }
-
-        std::unique_ptr<interface> clone() override {
-            return std::unique_ptr<interface>(new wrap<Impl>(wrapped));
-        }
-
-        Impl wrapped;
-    };
-};
-
-// Convenience routines for making schedule_generator:
 
 // Generate events with a fixed target and weight according to
 // a provided time schedule.
 
-struct schedule_generator {
-    schedule_generator(cell_local_label_type target, float weight, schedule sched):
+struct event_generator {
```

Comment:
In the PR message, you wrote the opposite: `event_generator` was suppressed in favor of a simplified `schedule_generator`. Accidental swap or am I missing sth?

Reply:
Grepping thought the docs finds mentions of `event_generator` that might need updates in:
1. `./cpp/hardware.rst`
2. `./cpp/recipe.rst`

Reply:
It's been a small dance
- rm event_generator (because that's the point)
- mv schedule_generator event_generator (since it's now the only kind of generator)
- So, now we still have event_generator, but it's different.

good point about the doc.

Reply:
Update the docs are fine; just callbacks returning an `event_generator`.